### PR TITLE
Simply exit DataLoader when Python is dying

### PR DESCRIPTION
**Pull request by ssnl**
I struggled with yet another DataLoader hang for the entire evening. After numerous experiments, I realized that it is unsafe to do anything when Python is shutting down. We also unfortunately implement our DataLaoder cleaning-up logic in `__del__`, a function that may or may not be called during shutdown, and if called, may or may not be called before core library resources are freed. 

Fortunately, we are already setting all our workers and pin_memory_thread as daemonic. So in case of Python shutting down, we can just do a no-op in `__del__` and rely on the automatic termination of daemonic children.

An `atexit` hook is used to detect Python exit.

### CODE REVIEW

**Comment by SsnL** [comment]
Also, the general recommendation is to never use `__del__` since it can't be relied upon. But I've yet to find a satisfying replacement.

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -320,21 +339,36 @@ class _DataLoaderIter(object):
     #      We delay the discussion on the logic in this case until later.
     #
     #   2. The iterator exits the workers when the loader process and/or worker
-    #      processes exits unexpectedly (e.g., SIGKILL-ed).
+    #      processes exits normally or with error.
```
Comment (SsnL) [inline torch/utils/data/dataloader.py]: This was worded wrongly before. The content is still correct.

**Review by ezyang — APPROVED** [review]
Thanks, love the commentary.

**Comment by SsnL** [comment]
last commit is only comment change and build passed on previous commit so I'm landing.

**Pull request closed without merging**